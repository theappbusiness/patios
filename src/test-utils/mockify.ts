export const mockify = <
  // TODO figure out the smart way to use Parameters<> to derive
  // the record values from typeof jest.fn so that we don't have
  // an any type here
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  TAnalog extends Record<string, (...args: any[]) => unknown>,
>(
  analog: TAnalog,
): TAnalog => {
  return Object.entries(analog).reduce<TAnalog>((acc, [key, value]) => {
    return {
      ...acc,
      [key]: jest.fn(value),
    }
  }, {} as TAnalog) as TAnalog
}
