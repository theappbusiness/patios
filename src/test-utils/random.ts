export const randomiseCase = (text: string): string =>
  text
    .split('')
    .map((char) =>
      Math.round(Math.random()) ? char.toUpperCase() : char.toLowerCase(),
    )
    .join('')

export const getRandomString = (): string =>
  randomiseCase(Math.random().toString(36).substring(2))
