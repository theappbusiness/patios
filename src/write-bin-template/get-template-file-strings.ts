import { stringify } from 'json-to-pretty-yaml'
import { FileSystem } from './types'

export type GetTemplateFileStrings = (jsonPath: string) => {
  jsonFileStr: string
  yamlFileStr: string
}

export const getTemplateFileStringsFactory =
  ({ readFileSync }: FileSystem) =>
  (jsonPath: string): { jsonFileStr: string; yamlFileStr: string } => {
    const rawTemplate = readFileSync(jsonPath).toString()
    const jsonFileStr = rawTemplate.replaceAll('$EXT', 'json')
    const yamlFileStr = stringify(
      JSON.parse(rawTemplate.replaceAll('$EXT', 'yaml')),
    )
    return {
      jsonFileStr,
      yamlFileStr,
    }
  }
