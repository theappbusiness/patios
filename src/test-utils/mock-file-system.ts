import type { FileSystem } from '../write-bin-template/types'
import { mockify } from './mockify'

// Feel free to fill in mkdirSync and rmSync if you think of something sensible
// eslint-disable-next-line @typescript-eslint/no-empty-function
const mkdirSync = (): void => {}
// eslint-disable-next-line @typescript-eslint/no-empty-function
const rmSync = (): void => {}

const readFileSync = (): Buffer => {
  return Buffer.from('Hello Files!')
}

const statSync = (dir: string): { isDirectory: () => boolean } => {
  switch (dir) {
    case '/home':
      return { isDirectory: () => true }
    case '/home/new_folder':
      return { isDirectory: () => false }
    default:
      return { isDirectory: () => false }
  }
}

const readdirSync = (dir: string): string[] => {
  switch (dir) {
    case '/home':
      return ['/a', '/b']
    case '/home/a':
      return ['/index.ts']
    case '/home/b':
      return ['/c']
    case '/home/b/c':
      return ['/deepA.mpg', '/deepB.jpeg']
    default:
      return []
  }
}

// eslint-disable-next-line @typescript-eslint/no-empty-function
const writeFileSync = (_path: string, _data: string): void => {}

export const makeMockFileSystem = (
  overrides?: Partial<FileSystem>,
): FileSystem => {
  const fileSystem: FileSystem = mockify({
    mkdirSync,
    readdirSync,
    readFileSync,
    rmSync,
    statSync,
    writeFileSync,
    ...overrides,
  })

  return fileSystem
}
