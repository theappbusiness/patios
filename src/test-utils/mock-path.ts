import { Path } from '../../types/node-stub.d'
import { mockify } from './mockify'

const dirname = (path: string): string => {
  const fileName = path.match(/.[a-z]{2,4}$/i)
  if (fileName && fileName.index) {
    return path.slice(0, fileName.index - 1)
  } else {
    return path
  }
}

const join = (...paths: string[]): string => {
  return paths.join('/')
}

const resolve = (dir: string, file: string): string => {
  return dir + file
}

export const makeMockPath = (overrides?: Partial<Path>): Path => {
  const path: Path = mockify({
    dirname,
    join,
    resolve,
    ...overrides,
  })

  return path
}
