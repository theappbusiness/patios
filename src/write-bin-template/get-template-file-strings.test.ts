import { makeMockFileSystem } from '../test-utils/mock-file-system'
import { getTemplateFileStringsFactory } from './get-template-file-strings'

const mockFile = {
  'just-a-string': "It's just words",
  'just-a-number': 8,
  an_array: ['someValue', { 'a nested': 'object' }],
  anObject: {
    anotherString: 'words inside an object',
    aReferenceToSomeExternalFile: '../wherever/it/is/referencedFileA.$EXT',
    anArray: [
      {
        some: 'thing',
        anotherReferenceToSomeExternalFile:
          '../wherever/it/is/referencedFileB.$EXT',
      },
    ],
  },
}

describe('getTemplateFileStrings', () => {
  let jsonFileStr: string, yamlFileStr: string
  beforeAll(() => {
    const mockFileSystem = makeMockFileSystem({
      readFileSync() {
        return Buffer.from(JSON.stringify(mockFile))
      },
    })
    const getTemplateFileStrings = getTemplateFileStringsFactory(mockFileSystem)
    ;({ jsonFileStr, yamlFileStr } = getTemplateFileStrings('/path'))
  })

  it('Converts a JSON template to stringified JSON', () => {
    expect(jsonFileStr).toMatchSnapshot()
  })

  it('Converts a JSON template to stringified YAML', () => {
    expect(yamlFileStr).toMatchSnapshot()
  })
})
