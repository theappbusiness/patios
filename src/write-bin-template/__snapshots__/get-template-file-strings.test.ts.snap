// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTemplateFileStrings Converts a JSON template to stringified JSON 1`] = `
"{
  "just-a-string": "It's just words",
  "just-a-number": 8,
  "an_array": [
    "someValue",
    {
      "a nested": "object"
    }
  ],
  "anObject": {
    "anotherString": "words inside an object",
    "aReferenceToSomeExternalFile": "../wherever/it/is/referencedFileA.json",
    "anArray": [
      {
        "some": "thing",
        "anotherReferenceToSomeExternalFile": "../wherever/it/is/referencedFileB.json"
      }
    ]
  }
}"
`;

exports[`getTemplateFileStrings Converts a JSON template to stringified YAML 1`] = `
"just-a-string: "It's just words"
just-a-number: 8
an_array:
  - "someValue"
  - a nested: "object"
anObject:
  anotherString: "words inside an object"
  aReferenceToSomeExternalFile: "../wherever/it/is/referencedFileA.yaml"
  anArray:
    - some: "thing"
      anotherReferenceToSomeExternalFile: "../wherever/it/is/referencedFileB.yaml"
"
`;
